// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

package cuopt.remote;

// Protocol version and metadata
message RequestHeader {
  uint32 version = 1;           // Protocol version (currently 1)
  ProblemType problem_type = 2; // LP or MIP
  IndexType index_type = 3;     // INT32 or INT64
  FloatType float_type = 4;     // FLOAT32 or DOUBLE
}

enum ProblemType {
  LP = 0;
  MIP = 1;
}

enum IndexType {
  INT32 = 0;
  INT64 = 1;
}

enum FloatType {
  FLOAT32 = 0;
  DOUBLE = 1;
}

// Optimization problem representation (field names match data_model_view_t)
message OptimizationProblem {
  // Problem metadata
  string problem_name = 1;
  string objective_name = 2;
  bool maximize = 3;
  double objective_scaling_factor = 4;
  double objective_offset = 5;
  // Field 6 reserved (was problem_category, now inferred from variable_types)

  // Variable and row names (optional)
  repeated string variable_names = 7;
  repeated string row_names = 8;

  // Constraint matrix A in CSR format (names match data_model_view_t: A_, A_indices_, A_offsets_)
  repeated double A = 10;
  repeated int32 A_indices = 11;
  repeated int32 A_offsets = 12;

  // Problem vectors (names match data_model_view_t: c_, b_)
  repeated double c = 20;              // objective coefficients
  repeated double b = 21;              // constraint bounds (RHS)
  repeated double variable_lower_bounds = 22;
  repeated double variable_upper_bounds = 23;

  // Constraint bounds (alternative to b + row_types)
  repeated double constraint_lower_bounds = 24;
  repeated double constraint_upper_bounds = 25;
  bytes row_types = 26;  // char array: 'E' (=), 'L' (<=), 'G' (>=), 'N' (objective)

  // Variable types (matches data_model_view_t: variable_types_)
  bytes variable_types = 30;  // char array: 'C' (continuous), 'I' (integer), 'B' (binary)

  // Initial solutions (matches data_model_view_t)
  repeated double initial_primal_solution = 40;
  repeated double initial_dual_solution = 41;

  // Quadratic objective matrix Q in CSR format for QPS (matches data_model_view_t: Q_objective_)
  repeated double Q_values = 50;
  repeated int32 Q_indices = 51;
  repeated int32 Q_offsets = 52;
}

// PDLP solver mode enum (matches cuOpt pdlp_solver_mode_t)
// Matches cuOpt pdlp_solver_mode_t enum values
enum PDLPSolverMode {
  Stable1 = 0;
  Stable2 = 1;
  Methodical1 = 2;
  Fast1 = 3;
  Stable3 = 4;
}

// Matches cuOpt method_t enum values
enum LPMethod {
  Concurrent = 0;
  PDLP = 1;
  DualSimplex = 2;
  Barrier = 3;
}

// PDLP solver settings (field names match cuOpt Python/C++ API)
message PDLPSolverSettings {
  // Termination tolerances
  double absolute_gap_tolerance = 1;
  double relative_gap_tolerance = 2;
  double primal_infeasible_tolerance = 3;
  double dual_infeasible_tolerance = 4;
  double absolute_dual_tolerance = 5;
  double relative_dual_tolerance = 6;
  double absolute_primal_tolerance = 7;
  double relative_primal_tolerance = 8;

  // Limits
  double time_limit = 10;
  // Iteration limit. Sentinel: set to -1 to mean "unset/use server defaults".
  // Note: proto3 numeric fields default to 0 when omitted, so clients should
  // explicitly use -1 (or a positive value) to avoid accidentally requesting 0 iterations.
  int64 iteration_limit = 11;

  // Solver configuration
  bool log_to_console = 20;
  bool detect_infeasibility = 21;
  bool strict_infeasibility = 22;
  PDLPSolverMode pdlp_solver_mode = 23;
  LPMethod method = 24;
  bool presolve = 25;
  bool dual_postsolve = 26;
  bool crossover = 27;
  int32 num_gpus = 28;

  // Advanced options
  bool per_constraint_residual = 30;
  bool cudss_deterministic = 31;
  int32 folding = 32;
  int32 augmented = 33;
  int32 dualize = 34;
  int32 ordering = 35;
  int32 barrier_dual_initial_point = 36;
  bool eliminate_dense_columns = 37;
  bool save_best_primal_so_far = 38;
  bool first_primal_feasible = 39;

  // Warm start data (if provided)
  PDLPWarmStartData warm_start_data = 50;
}

message PDLPWarmStartData {
  repeated double current_primal_solution = 1;
  repeated double current_dual_solution = 2;
  repeated double initial_primal_average = 3;
  repeated double initial_dual_average = 4;
  repeated double current_ATY = 5;
  repeated double sum_primal_solutions = 6;
  repeated double sum_dual_solutions = 7;
  repeated double last_restart_duality_gap_primal_solution = 8;
  repeated double last_restart_duality_gap_dual_solution = 9;

  double initial_primal_weight = 10;
  double initial_step_size = 11;
  int32 total_pdlp_iterations = 12;
  int32 total_pdhg_iterations = 13;
  double last_candidate_kkt_score = 14;
  double last_restart_kkt_score = 15;
  double sum_solution_weight = 16;
  int32 iterations_since_last_restart = 17;
}

// MIP solver settings (field names match cuOpt Python/C++ API)
message MIPSolverSettings {
  // Limits
  double time_limit = 1;

  // Tolerances
  double relative_mip_gap = 2;
  double absolute_mip_gap = 3;
  double integrality_tolerance = 4;
  double absolute_tolerance = 5;
  double relative_tolerance = 6;
  double presolve_absolute_tolerance = 7;

  // Solver configuration
  bool log_to_console = 10;
  bool heuristics_only = 11;
  int32 num_cpu_threads = 12;
  int32 num_gpus = 13;
  bool presolve = 14;
  bool mip_scaling = 15;
}

// LP solve request
message SolveLPRequest {
  RequestHeader header = 1;
  OptimizationProblem problem = 2;
  PDLPSolverSettings settings = 3;
}

// MIP solve request
message SolveMIPRequest {
  RequestHeader header = 1;
  OptimizationProblem problem = 2;
  MIPSolverSettings settings = 3;
  optional bool enable_incumbents = 4;
}

// LP solution
message LPSolution {
  // Solution vectors
  repeated double primal_solution = 1;
  repeated double dual_solution = 2;
  repeated double reduced_cost = 3;

  // Warm start data for next solve
  PDLPWarmStartData warm_start_data = 4;

  // Termination information
  PDLPTerminationStatus termination_status = 10;
  string error_message = 11;

  // Solution statistics
  double l2_primal_residual = 20;
  double l2_dual_residual = 21;
  double primal_objective = 22;
  double dual_objective = 23;
  double gap = 24;
  int32 nb_iterations = 25;
  double solve_time = 26;
  bool solved_by_pdlp = 27;
}

enum PDLPTerminationStatus {
  PDLP_NO_TERMINATION = 0;
  PDLP_NUMERICAL_ERROR = 1;
  PDLP_OPTIMAL = 2;
  PDLP_PRIMAL_INFEASIBLE = 3;
  PDLP_DUAL_INFEASIBLE = 4;
  PDLP_ITERATION_LIMIT = 5;
  PDLP_TIME_LIMIT = 6;
  PDLP_CONCURRENT_LIMIT = 7;
  PDLP_PRIMAL_FEASIBLE = 8;
}

// MIP solution
message MIPSolution {
  repeated double solution = 1;

  MIPTerminationStatus termination_status = 10;
  string error_message = 11;

  double objective = 20;
  double mip_gap = 21;
  double solution_bound = 22;
  double total_solve_time = 23;
  double presolve_time = 24;
  double max_constraint_violation = 25;
  double max_int_violation = 26;
  double max_variable_bound_violation = 27;
  int32 nodes = 28;
  int32 simplex_iterations = 29;
}

enum MIPTerminationStatus {
  MIP_NO_TERMINATION = 0;
  MIP_OPTIMAL = 1;
  MIP_FEASIBLE_FOUND = 2;
  MIP_INFEASIBLE = 3;
  MIP_UNBOUNDED = 4;
  MIP_TIME_LIMIT = 5;
}

// Job status for async operations
enum JobStatus {
  QUEUED = 0;           // Job submitted, waiting in queue
  PROCESSING = 1;       // Job currently being solved
  COMPLETED = 2;        // Job completed successfully
  FAILED = 3;           // Job failed with error
  NOT_FOUND = 4;        // Job ID not found
  CANCELLED = 5;        // Job was cancelled by user
}

// Response for job submission
message SubmitResponse {
  ResponseStatus status = 1;
  bytes job_id = 2;           // Unique job identifier (bytes to avoid UTF-8 validation warnings)
  string message = 3;         // Success/error message
}

// Response for status check
message StatusResponse {
  JobStatus job_status = 1;
  string message = 2;
  double progress = 3;        // 0.0-1.0 (future enhancement)
  int64 result_size_bytes = 4;    // Size of result payload when COMPLETED (0 if unknown)
  int64 max_message_bytes = 5;    // Server gRPC max message size (-1 = unlimited)
}

// Response for get result
message ResultResponse {
  ResponseStatus status = 1;
  string error_message = 2;

  oneof solution {
    LPSolution lp_solution = 10;
    MIPSolution mip_solution = 11;
  }
}

// Response for delete
message DeleteResponse {
  ResponseStatus status = 1;
  string message = 2;
}

// Response for cancel job
message CancelResponse {
  ResponseStatus status = 1;
  string message = 2;
  JobStatus job_status = 3;       // Status of job after cancel attempt
}

enum ResponseStatus {
  SUCCESS = 0;
  ERROR_INVALID_REQUEST = 1;
  ERROR_SOLVE_FAILED = 2;
  ERROR_INTERNAL = 3;
  ERROR_NOT_FOUND = 4;
}
